cmd/restic/cmd_prune.go

```diff
@@ -56,207 +101,348 @@ func runPrune(gopts GlobalOptions) error {
 	}
 
-	Verbosef("will delete %d packs and rewrite %d packs, this frees %s\n",
-		len(removePacks), len(rewritePacks), formatBytes(uint64(removeBytes)))
+	// If unused space is restricted, check and find packs for repacking
+	if opts.MaxUnusedPercent < 100.0 {
+		maxUnusedSizeAfter := uint64(opts.MaxUnusedPercent / (100.0 - opts.MaxUnusedPercent) * float32(stats.size.used))
 
-	var obsoletePacks restic.IDSet
-	if len(rewritePacks) != 0 {
-		bar := newProgressMax(!gopts.Quiet, uint64(len(rewritePacks)), "packs rewritten")
-		obsoletePacks, err = repository.Repack(ctx, repo, rewritePacks, usedBlobs, bar)
-		if err != nil {
-			return err
+		//sort repackCandidates such that packs with highest ratio unused/used space are picked first
+		sort.Slice(repackCandidates, func(i, j int) bool {
+			return repackCandidates[i].unusedSize*repackCandidates[j].usedSize >
+				repackCandidates[j].unusedSize*repackCandidates[i].usedSize
+		})
+
+		for i, p := range repackCandidates {
+			if stats.size.unused < maxUnusedSizeAfter+stats.size.remove+stats.size.repackrm || uint(len(repackPacks)) >= opts.MaxRepackCount {
+				stats.packs.keep += uint(len(repackCandidates) - i)
+				break
+			}
+			repack(p.ID, p.packInfo)
 		}
 	}
 
-	removePacks.Merge(obsoletePacks)
+	Verbosef("\nused:        %10d blobs / %s\n", stats.blobs.used, formatBytes(stats.size.used))
+	Verbosef("duplicates:  %10d blobs / %s\n", stats.blobs.duplicate, formatBytes(stats.size.duplicate))
+	Verbosef("unused:      %10d blobs / %s\n", stats.blobs.unused, formatBytes(stats.size.unused))
+	Verbosef("unreferenced:                   %s\n", formatBytes(stats.size.unref))
+	totalSize := stats.size.used + stats.size.duplicate + stats.size.unused + stats.size.unref
+	Verbosef("total:       %10d blobs / %s\n", stats.blobs.used+stats.blobs.unused+stats.blobs.duplicate,
+		formatBytes(totalSize))
+	Verbosef("unused size: %s of total size\n\n", formatPercent(stats.size.unused, totalSize))
+
+	Verbosef("to repack:   %10d blobs / %s\n", stats.blobs.repack, formatBytes(stats.size.repack))
+	Verbosef("  -> prunes: %10d blobs / %s\n", stats.blobs.repackrm, formatBytes(stats.size.repackrm))
+	Verbosef("to delete:   %10d blobs / %s\n", stats.blobs.remove, formatBytes(stats.size.remove))
+	Verbosef("delete unreferenced:            %s\n", formatBytes(stats.size.unref))
+	totalPruneSize := stats.size.remove + stats.size.repackrm + stats.size.unref
+	Verbosef("total prune: %10d blobs / %s\n", stats.blobs.remove+stats.blobs.repackrm, formatBytes(totalPruneSize))
+	Verbosef("unused size after prune: %s of total size\n\n",
+		formatPercent(stats.size.unused-stats.size.remove-stats.size.repackrm, totalSize-totalPruneSize))
+
+	Verbosef("mixed packs:        %10d\n", stats.packs.mixed)
+	Verbosef("tree packs:         %10d\n", stats.packs.tree)
+	Verbosef("data packs:         %10d\n", stats.packs.data)
+	Verbosef("unreferenced packs: %10d\n\n", len(removePacksFirst))
+	Verbosef("total packs:        %10d\n\n",
```

Comment:
I removed this.

Actually I just realized that if we want to print out those statistics, it would be even better to just print out what is kept/repacked for what reason, like
```
keeping because of --max-repack-count: x packs
repacking because of --repack-mixed: x packs
repacking because of --repack-small: x packs
```
and so on..

Anyway, I still have to work on verbosity and statistics...

Reply:
`repacking x mixed packs` etc. would sound more natural.

Reply:
I just removed some statistics and show some only with more verbosity. If we need more/other statistics in future, I can add them in another PR.